MoltenVK/MoltenVK/GPUObjects/MVKImage.mm

```diff
@@ -1190,12 +1190,16 @@
 									 uint32_t swapchainIndex) : MVKImage(device, pCreateInfo) {
 	_swapchain = swapchain;
 	_swapchainIndex = swapchainIndex;
+}
 
-	_swapchain->retain();
+void MVKSwapchainImage::detachSwapchain() {
+	lock_guard<mutex> lock(_swapchainLock);
+	_swapchain = nullptr;
 }
 
-MVKSwapchainImage::~MVKSwapchainImage() {
-	if (_swapchain) { _swapchain->release(); }
+void MVKSwapchainImage::destroy() {
+	detachSwapchain();
```

Comment:
This is the only call site for `detachSwapchain()`. Is that intentional? The code in `presentCAMetalDrawable()` implies that it was supposed to be called elsewhere, e.g. when the swapchain is destroyed. Otherwise, I fail to see how the swapchain pointer becomes `nullptr` after the swapchain is destroyed.

Reply:
For each swap chain image, `MVKSwapchain::destroy()` calls `MVKPresentableSwapchainImage::destroy()` (via `_device->destroyPresentableSwapchainImage()`).

So the app request to destroy the swapchain triggers the `detachSwapchain()` and breaks the link back to the swapchain from the image. The swapchain itself is not retained by the images, so it is free to deallocate at that time.

In the end, the drawable-presented callback block will hold onto the swapchain image, but not the swapchain itself once it has been destroyed by the app. When that callback block executes, it will find the swapchain gone, and won't record the timings.

Reply:
> For each swap chain image, `MVKSwapchain::destroy()` calls `MVKPresentableSwapchainImage::destroy()` (via `_device->destroyPresentableSwapchainImage()`).
> ...
> In the end, the drawable-presented callback block will hold onto the swapchain image, but not the swapchain itself once it has been destroyed by the app. When that callback block executes, it will find the swapchain gone, and won't record the timings.

All right, I see now.